### ISSUE TITLE
rpi black screen in all windows.

### ISSUE THREAD

**Issue by GWHAYWOOD** [Xpra-org/xpra#3288]
xpra version 4.3-r867 on the server, which is an Intel i7 laptop.
xpra version 4.3-r29977 (g8a0d2b663) on the client which is a Raspberry Pi 3B+.

I log in via ssh from the client to the server, where there is no window manager running and no other user session running.

I type
```shell
$ xpra start :103
$ DISPLAY=:103 xterm
```
to start a bash shell in an xterm on the server using xpra as the window manager.

On the client from which I log in to the server (on the client there IS a window manager running - fluxbox) I type
```
xpra attach ssh://laptop3/103
```
which after a little time opens a window on the client with a completely black display although the window ornamentation is as expected.  I can use this window to type commands, but there is no visible indication on the black screen of any response.  The commands are however executed on the server, for example I can 'touch' a non-existent file to create it in the user's home directory on the server.

The results are exactly the same if I use `--opengl=yes` or `--opengl=no`.

The bug-report data included a screenshot which I have removed.  It was an image of the "file manager" display of the user's home directory.  It contained sensitive information and nothing which could possibly be of use in diagnosing this issue.  I have attached a tarball of the directory excluding the screenshot.  That is the only redaction, but it was not clear to me that the bug-report process completed successfully.  I used the 'save' button to create the directory of report contents but then had to kill the bug-report process as there seemed to be no obvious way to complete it.

The attached server log file is for three attempts to connect from the client:
```
xpra attach ssh://laptop3/103
xpra --opengl=yes  attach ssh://laptop3/103
xpra --opengl=no attach ssh://laptop3/103
```
[XPRA_bug_report_data.tar.gz](https://github.com/Xpra-org/xpra/files/7245120/XPRA_bug_report_data.tar.gz)
[xpra.log](https://github.com/Xpra-org/xpra/files/7245309/xpra.log)

**Comment by totaam** [comment]
Original mailing list thread: https://lists.devloop.org.uk/pipermail/shifter-users/2021-September/002876.html

> ```
> xpra start :103
> DISPLAY=:103 xterm
> ```

FYI: always use `xpra start --start=xterm` instead.

There is no mention of `/run/user/$UID/xpra` anywhere, which is strange.
It does write to `/tmp/xpra/` instead. I guess this server was started from an ssh session?

Other information gleaned from the bug report data and log file.
* server is `xpra X11 seamless version 4.3-r867 64-bit` `running with pid 21107 on Linux Debian 10 buster`
* client is `Python/GTK3 Linux Raspbian 11 bullseye tty client version 4.3-r29977` - odd, what does `tty` mean here?
* the only encoders supported are the builtin (`rgb` and `scroll`) and the pillow ones (`png*` and `jpeg`)

How was this installed? From source?

> The bug-report data included a screenshot which I have removed

You can just tell the tool not to take the screenshot. Though it is extremely useful, especially when the bug is a visual one.
A better way would be to move your sensitive applications out of the way before taking the screenshot.

Please try starting and attaching to a local session on the server and on the client separately:
`xpra start --start=xterm --attach=yes`
To see if the problem still occurs.
Or you could also try connecting with a different client (ie: MS Windows) or install `xpra-html5` and use a browser.
Then also try turning off transparency support, both on the server and on the client:
```
XPRA_ALPHA=0 xpra start ...
```

You could also try forcing encodings: `--encodings=png`.

**Comment by GWHAYWOOD** [comment]
On Wed, 29 Sep 2021, Antoine Martin wrote:

> Original mailing list thread: https://lists.devloop.org.uk/pipermail/shifter-users/2021-September/002876.html
>
>> ```
>> xpra start :103
>> DISPLAY=:103 xterm
>> ```
>
> FYI: always use `xpra start --start=xterm` instead.

Same result.

> There is no mention of `/run/user/$UID/xpra` anywhere, which is strange.

It is mentioned, but on stdout:

8<----------------------------------------------------------------------
laptop3:~$ >>> xpra start --start=xterm :103
Warning: XDG_RUNTIME_DIR is not defined
  and '/run/user/1001' does not exist
  using '/tmp'
laptop3:~$ >>> Entering daemon mode; any further errors will be reported to:
   /tmp/xpra/103/server.log
8<----------------------------------------------------------------------

> It does write to `/tmp/xpra/` instead. I guess this server was started from an ssh session?

Correct.

> Other information gleaned from the bug report data and log file.
> * server is `xpra X11 seamless version 4.3-r867 64-bit` `running with pid 21107 on Linux Debian 10 buster`
> * client is `Python/GTK3 Linux Raspbian 11 bullseye tty client version 4.3-r29977` - odd, what does `tty` mean here?

This may or may not be a clue - I have no idea - but it reminds me of
something which I noticed only last week.  Frequently I run 'cal' to
look at dates.  Most often when I do that I'm logged into what for the
purposes of this discussion is the xpra server.  Most often, the login
session is a remote X session but it may be an ordinary 'ssh' session.
Because, since the latest Debian 'upgrades', it has been so painfully
slow to use X remotely, I have taken to running 'cal' on the local box
(the same Raspberry Pi which is giving the black xterm display when I
use xpra attach).  On the server, both when running X/fvwm2 *and* when
logging in via 'ssh', if 'cal' displays the current date, then this is
highlighted in reverse video.  On the Pi, 'cal' does *not* display the
current date in reverse video.  I wondered why not; the 'man' page for
'cal' says that it is not highlighted if the terminal is not a tty!

Although inconvenient, it didn't seem to be a pressing issue so I have
not investigated.  Perhaps now I will.

> * the only encoders supported are the builtin (`rgb` and `scroll`) and the pillow ones (`png*` and `jpeg`)

At this point I should make clear that I don't know what encoders are,
nor what they are for (although I can hazard a guess) nor how they are
installed, nor if I should install others, and if so how I should both
choose and install them.  But I'm happy to do it, if it might help. :)

> How was this installed? From source?

Yes, from source, using 'git clone' - because I read what you had to
say about distro packages (and I couldn't agree more).  See my OP at

https://lists.devloop.org.uk/pipermail/shifter-users/2021-September/002870.html

>> The bug-report data included a screenshot which I have removed
>
> You can just tell the tool not to take the screenshot. Though it is extremely useful, especially when the bug is a visual one.
> A better way would be to move your sensitive applications out of the way before taking the screenshot.
>
> Please try starting and attaching to a local session on the server and on the client separately:

As I explained in my OP, this works fine on the server but gives the
same symptoms on the client.

> `xpra start --start=xterm --attach=yes`
> To see if the problem still occurs.

You need to be aware that I am completely unfamiliar with xpra.  I do
not understand the instruction, but I have done my best to follow it.
However in both cases I need to have an X server and window manager
running to get any result - otherwise I get no response at all other
than the brief message to stdout about logging.  To recap:

xpra start --start=xterm --attach=yes :109

given in an xterm using the 'fvwm' window manager on the server gives
a fully functioning xterm running 'bash' on display 109 as I expect.

On the client, using the 'fluxbox' window manager, as usual I get a
functional xterm (I can type 'blind' and the commands are executed)
with the proper border ornamentation but a completely black window:

8<----------------------------------------------------------------------
piplus:~$ >>> xpra start --start=xterm --attach=yes :110
You have mail in /var/mail/ged
piplus:~$ >>> Entering daemon mode; any further errors will be reported to:
   /run/user/1001/xpra/110/server.log
8<----------------------------------------------------------------------

server.log:

8<----------------------------------------------------------------------
failure on /tmp/.X11-unix/X110: [Errno 2] No such file or directory: '/tmp/.X11-unix/X110'
2021-09-29 10:44:35,450 cannot access python uinput module:[0m
2021-09-29 10:44:35,451  No module named 'uinput'[0m
[31m2021-09-29 10:44:37,960 dbus-launch failed to start using command 'dbus-launch --sh-syntax --close-stderr':
[0m
[31m2021-09-29 10:44:37,961  [Errno 2] No such file or directory: 'dbus-launch'
[0m
2021-09-29 10:44:38,446 pointer device emulation using XTest[0m
[31m2021-09-29 10:44:38,706 Error: cannot find the html web root[0m
[31m2021-09-29 10:44:38,706  '/usr/local/share/xpra/www' does not exist[0m
[31m2021-09-29 10:44:38,707  install the html-xpra package[0m
2021-09-29 10:44:38,708 wrote pid 10311 to '/run/user/1001/xpra/110/server.pid'[0m
2021-09-29 10:44:38,800 created unix domain socket '/run/user/1001/xpra/piplus-110'[0m
2021-09-29 10:44:38,801 cannot create group socket '/run/xpra/piplus-110'[0m
2021-09-29 10:44:38,801  [Errno 13] Permission denied[0m
2021-09-29 10:44:38,803 created unix domain socket '/home/ged/.xpra/piplus-110'[0m
2021-09-29 10:44:38,804 xvfb pid=10326[0m
2021-09-29 10:44:39,111 D-Bus notification forwarding is available[0m
2021-09-29 10:44:39,171 pulseaudio server started with pid 10371[0m
2021-09-29 10:44:39,172  private server socket path:[0m
2021-09-29 10:44:39,172  '/run/user/1001/xpra/110/pulse-110/pulse/native'[0m
W: [pulseaudio] server-lookup.c: Unable to contact D-Bus: org.freedesktop.DBus.Error.Spawn.ExecFailed: /usr/bin/dbus-launch terminated abnormally without any error message
W: [pulseaudio] main.c: Unable to contact D-Bus: org.freedesktop.DBus.Error.Spawn.ExecFailed: /usr/bin/dbus-launch terminated abnormally without any error message
2021-09-29 10:44:39,671 watching for applications menu changes in:[0m
2021-09-29 10:44:39,672  '/usr/share/applications'[0m
2021-09-29 10:44:39,673  '/usr/local/share/applications'[0m
[33m2021-09-29 10:44:39,817 Warning: webcam forwarding is disabled[0m
[33m2021-09-29 10:44:39,818  the virtual video directory '/sys/devices/virtual/video4linux' was not found[0m
[33m2021-09-29 10:44:39,818  make sure that the 'v4l2loopback' kernel module is installed and loaded[0m
[33m2021-09-29 10:44:39,818  or use the 'webcam=no' option[0m
2021-09-29 10:44:39,819 found 0 virtual video devices for webcam forwarding[0m
2021-09-29 10:44:39,878 printer forwarding enabled using postscript and pdf[0m
2021-09-29 10:44:39,901 started command '/etc/X11/Xsession true' with pid 10383[0m
[31m2021-09-29 10:44:39,929 Error parsing xdg menu data:[0m
[31m2021-09-29 10:44:39,929  ParsingError in file '/etc/xdg/menus/kde-debian-menu.menu', File not found[0m
[31m2021-09-29 10:44:39,931  this is either a bug in python-xdg,[0m
[31m2021-09-29 10:44:39,931  or an invalid system menu configuration[0m
2021-09-29 10:44:39,933 started command 'xterm' with pid 10387[0m
Warning: failed to import GStreamer 1.x:
  Namespace Gst not available
[31m2021-09-29 10:44:40,236 Error: failed to query sound subsystem:[0m
[31m2021-09-29 10:44:40,238  query did not return any data[0m

(Xpra:10311): Gtk-CRITICAL **: 10:44:40.244: gtk_widget_realize: assertion 'widget->priv->anchored || GTK_IS_INVISIBLE (widget)' failed
2021-09-29 10:44:40,291 0.9GB of system memory[0m
2021-09-29 10:44:41,556 waiting for initialization thread to complete[0m
2021-09-29 10:44:42,492 xpra is ready.[0m
2021-09-29 10:44:42,505 xpra X11 seamless version 4.3-r29977 (g8a0d2b663) 32-bit[0m
2021-09-29 10:44:42,507  uid=1001 (ged), gid=1001 (ged)[0m
2021-09-29 10:44:42,567  running with pid 10311 on Linux Raspbian 11 bullseye[0m
2021-09-29 10:44:43,263  connected to X11 display :110 with 24 bit colors[0m
2021-09-29 10:44:46,108 Xpra GTK3 client version 4.3-r29977 (g8a0d2b663) 32-bit
2021-09-29 10:44:46,694  running on Linux Raspbian 11 bullseye
2021-09-29 10:44:46,697  window manager is 'Fluxbox'
[33m2021-09-29 10:44:51,557 Warning: initialization thread is still active[0m
Warning: failed to import GStreamer 1.x:
  Namespace Gst not available
2021-09-29 10:44:52,987 Error: failed to query sound subsystem:
2021-09-29 10:44:52,988  query did not return any data
2021-09-29 10:44:53,241 OpenGL is supported on display ':110'[0m
2021-09-29 10:44:53,242  using 'llvmpipe (LLVM 11.0.1, 128 bits)' renderer[0m
2021-09-29 10:44:53,340 Warning: failed to load the GTK3 notification class
2021-09-29 10:44:53,341  Namespace Notify not available
2021-09-29 10:44:54,888 created unix domain socket '/run/user/1001/xpra/clients/piplus-10412'
2021-09-29 10:44:55,269 No OpenGL_accelerate module loaded: No module named 'OpenGL_accelerate'
2021-09-29 10:44:56,384 Warning: cannot handle window transparency
2021-09-29 10:44:56,385  screen is not composited
2021-09-29 10:44:56,863 Warning: window 0xffffffff changed its transparency attribute
2021-09-29 10:44:56,864  from False to True, behaviour is undefined
2021-09-29 10:44:56,964 OpenGL enabled with llvmpipe (LLVM 11.0.1, 128 bits)
2021-09-29 10:44:57,785  keyboard settings: rules=evdev, model=pc105, layout=gb
2021-09-29 10:44:57,799  desktop size is 1600x900:
2021-09-29 10:44:57,801   :0.0 (423x238 mm - DPI: 96x96) workarea: 1600x879 
2021-09-29 10:44:58,014 New unix-domain connection received[0m
2021-09-29 10:44:58,014  on '/run/user/1001/xpra/piplus-110'[0m
2021-09-29 10:44:58,037 Handshake complete; enabling connection[0m
2021-09-29 10:44:58,173  mmap is enabled using 256MB area in /run/user/1001/xpra/xpra.au1_ciep.mmap[0m
2021-09-29 10:44:58,178 Python/GTK3 Linux Raspbian 11 bullseye tty client version 4.3-r29977 32-bit[0m
2021-09-29 10:44:58,178  OpenGL is enabled with llvmpipe (LLVM 11.0.1, 128 bits)[0m
2021-09-29 10:44:58,178  connected from 'piplus' as 'ged' - 'Ged'[0m
2021-09-29 10:44:58,207 setting key repeat rate from client: 660ms delay / 40ms interval[0m
2021-09-29 10:44:58,213 setting keymap: rules=evdev, model=pc105, layout=gb[0m
2021-09-29 10:44:58,556 setting keyboard layout to 'gb'[0m
2021-09-29 10:44:59,069  client root window size is 1600x900 with 1 display:[0m
2021-09-29 10:44:59,071   :0.0 (423x238 mm - DPI: 96x96) workarea: 1600x879 [0m
2021-09-29 10:44:59,619 server virtual display now set to 1600x900[0m
2021-09-29 10:44:59,633  automatic picture encoding enabled, also available:[0m
2021-09-29 10:44:59,634   png, png/P, png/L, rgb24, rgb32, jpeg, scroll[0m
[31m2021-09-29 10:44:59,682 Error parsing xdg menu data:[0m
[31m2021-09-29 10:44:59,683  ParsingError in file '/etc/xdg/menus/kde-debian-menu.menu', File not found[0m
[31m2021-09-29 10:44:59,683  this is either a bug in python-xdg,[0m
[31m2021-09-29 10:44:59,683  or an invalid system menu configuration[0m
2021-09-29 10:44:59,722 enabled fast mmap transfers using 256MB shared memory area
2021-09-29 10:44:59,724 enabled remote logging
2021-09-29 10:44:59,726 Xpra X11 seamless server version 4.3-r29977 32-bit
2021-09-29 10:44:59,727  running on Linux Raspbian 11 bullseye
2021-09-29 10:44:59,728 client   1 @13.612 Xpra X11 seamless server version 4.3-r29977 32-bit[0m
2021-09-29 10:44:59,731 client   1 @13.614  running on Linux Raspbian 11 bullseye[0m
2021-09-29 10:44:59,765 Attached to socket:///run/user/1001/xpra/piplus-110
2021-09-29 10:44:59,766  (press Control-C to detach)

2021-09-29 10:44:59,783 client   1 @13.652 Attached to socket:///run/user/1001/xpra/piplus-110[0m
2021-09-29 10:44:59,788 client   1 @13.653  (press Control-C to detach)[0m
2021-09-29 10:45:00,069 DPI set to 96 x 96[0m
TypeError: Couldn't find foreign struct converter for 'cairo.Context'
2021-09-29 10:45:00,228 running, 1 window
2021-09-29 10:45:00,232 client   1 @14.115 running, 1 window[0m
TypeError: Couldn't find foreign struct converter for 'cairo.Context'
TypeError: Couldn't find foreign struct converter for 'cairo.Context'
TypeError: Couldn't find foreign struct converter for 'cairo.Context'
[31m2021-09-29 10:45:00,624 Error parsing xdg menu data:[0m
[31m2021-09-29 10:45:00,625  ParsingError in file '/etc/xdg/menus/kde-debian-menu.menu', File not found[0m
[31m2021-09-29 10:45:00,626  this is either a bug in python-xdg,[0m
[31m2021-09-29 10:45:00,627  or an invalid system menu configuration[0m
2021-09-29 10:45:02,196 Error: cannot access the list of printers
[31m2021-09-29 10:45:02,199 client   1 @16.083 Error: cannot access the list of printers[0m
2021-09-29 10:45:02,199  module 'cups' has no attribute 'Connection'
[31m2021-09-29 10:45:02,202 client   1 @16.085  module 'cups' has no attribute 'Connection'[0m
TypeError: Couldn't find foreign struct converter for 'cairo.Context'
/usr/lib/python3/dist-packages/gi/overrides/Gtk.py:1649: Warning: g_object_ref: assertion 'G_IS_OBJECT (object)' failed
   return _Gtk_main(*args, **kwargs)
8<----------------------------------------------------------------------

> Or you could also try connecting with a different client (ie: MS
> Windows) or install `xpra-html5` and use a browser.

That will take a little time to set up but I will try it and report.

> Then also try turning off transparency support, both on the server and on the client:
> ```
> XPRA_ALPHA=0 xpra start ...
> ```

This gives the same result, a completely black window content.

> You could also try forcing encodings: `--encodings=png`.

When I use this option, no xterm appears at all.

Thank you for the suggestions.  As I said I will report on xpra-html5,
although I should be most reluctant to use a browser for this purpose
routinely.  Running a browser on the Pi3B+ is unsatisfactory at best,
and sometimes crashes it.  This is one reason for wanting to use xpra
in the first place.

-- 

73,
Ged.

**Comment by totaam** [comment]
>> There is no mention of `/run/user/$UID/xpra` anywhere, which is strange.
>>
> It is mentioned, but on stdout:
>
Right, that's on Debian.
It's not ideal to run without a valid `$XDG_RUNTIME_DIR` but it should still work.

> 'cal' says that it is not highlighted if the terminal is not a tty

This is unlikely to be related, but who knows.

> At this point I should make clear that I don't know what encoders are...

Some related information can be found here:
https://github.com/Xpra-org/xpra/blob/master/docs/Usage/Encodings.md

> As I explained in my OP, this works fine on the server but gives the same symptoms on the client.

OK, so definitely something weird happening with the client then.
I was hoping that running both server and client on that system would resolve things. Either because the client and server can enable `mmap` when running locally, or because the (potentially weird) pixel format order wouldn't matter.

> `XPRA_ALPHA=0 xpra start ...`
> When I use this option, no xterm appears at all.

Damn.

> As I said I will report on xpra-html5, although I should be most reluctant to use a browser for this purpose routinely.

I am not suggesting that you should use the browser routinely, only for diagnostics.
This should give us an interesting data point.

I have located my rpi, now I need to put an OS on it and wire it up somewhere.

**Comment by GWHAYWOOD** [comment]
On Wed, 29 Sep 2021, Antoine Martin wrote:

> It's not ideal to run without a valid `$XDG_RUNTIME_DIR` but it should still work.

I tried creating one.  No change to the symptoms.

> https://github.com/Xpra-org/xpra/blob/master/docs/Usage/Encodings.md

Thanks.  A little clearer.

On Wed, 29 Sep 2021, G.W. Haywood wrote:

>> As I said I will report on xpra-html5, although I should be most
>> reluctant to use a browser for this purpose routinely.
> 
> I am not suggesting that you should use the browser routinely, only
> for diagnostics.

I understand.

> This should give us an interesting data point.

Not much luck I'm afraid.  I didn't even get as far with the HTML
client as I did with 'xpra attach'.  The browser renders a page
containing a progress bar.  The progress indicator stops before
reaching its end, usually at about half-way but sometimes at 10%
approximately.  Then it just sits there indefinitely.  Using the
'telnet' utility I verified that I can connect to the server, so
the server is definitely listening and the browser seems to be
able to connect to the server.

In desperation I reinstalled a few things including OpenGL_Accelerate,
and redownloaded (git clone), rebuilt and reinstalled xpra on the Pi.
There's an OpenGL error message on stdout which I don't remember seeing
before.  Here's stdout on the Pi:

8<----------------------------------------------------------------------
piplus:~$ >>> xpra attach ssh://laptop3/191
2021-09-30 13:01:52,466 Xpra GTK3 client version 4.3-r30103 (gc787b6465) 32-bit
2021-09-30 13:01:52,938  running on Linux Raspbian 11 bullseye
2021-09-30 13:01:52,942  window manager is 'Fluxbox'
Warning: failed to import GStreamer 1.x:
  Namespace Gst not available
2021-09-30 13:01:59,649 Error: failed to query sound subsystem:
2021-09-30 13:01:59,650  query did not return any data
2021-09-30 13:02:00,698 Warning: failed to load the GTK3 notification class
2021-09-30 13:02:00,698  Namespace Notify not available
2021-09-30 13:02:00,765 failed to instantiate the dbus notification handler:
2021-09-30 13:02:00,767  you may need to start a notification service for 'org.freedesktop.Notifications'
2021-09-30 13:02:00,768  disable notifications to avoid this warning
2021-09-30 13:02:01,962 created unix domain socket '/run/user/1001/xpra/clients/piplus-1206'
2021-09-30 13:02:02,553 OpenGL_accelerate module loaded
2021-09-30 13:02:02,579 Using accelerated ArrayDatatype
2021-09-30 13:02:04,450 Warning: cannot handle window transparency
2021-09-30 13:02:04,450  screen is not composited
2021-09-30 13:02:06,056 Error loading OpenGL support:
2021-09-30 13:02:06,057  ('numpy.ndarray size changed, may indicate binary incompatibility. Expected 44 from C header, got 40 from PyObject', 1, <OpenGL.platform.baseplatform.glGetIntegerv object at 0x72a72408>)
2021-09-30 13:02:06,802 Connected (version 2.0, client OpenSSH_7.9p1)
2021-09-30 13:02:06,986 Authentication (publickey) successful!
2021-09-30 13:02:07,101  keyboard settings: rules=evdev, model=pc105, layout=gb
2021-09-30 13:02:07,118  desktop size is 1600x900:
2021-09-30 13:02:07,121   :0.0 (423x238 mm - DPI: 96x96) workarea: 1600x879 
2021-09-30 13:02:07,548  SSH: 'Warning: XDG_RUNTIME_DIR is not defined'
2021-09-30 13:02:07,549  SSH: " and '/run/user/1001' does not exist"
2021-09-30 13:02:07,553  SSH: " using '/tmp'"
2021-09-30 13:02:07,680 enabled remote logging
2021-09-30 13:02:07,683 Xpra X11 seamless server version 4.3-r867 64-bit
2021-09-30 13:02:07,687  running on Linux Debian 10 buster
2021-09-30 13:02:07,744 Attached to xpra server at ***@***.***/191
2021-09-30 13:02:07,745  (press Control-C to detach)

2021-09-30 13:02:08,959 running
2021-09-30 13:02:10,205 Error: cannot access the list of printers
2021-09-30 13:02:10,206  module 'cups' has no attribute 'Connection'
^[[A^[[B^C
2021-09-30 13:02:52,181 Python/GTK3 Client got signal SIGINT
2021-09-30 13:02:52,183 exiting
2021-09-30 13:02:52,205 SSH EOF on stderr of run-xpra
2021-09-30 13:02:52,388 removing unix domain socket '/run/user/1001/xpra/clients/piplus-1206'
8<----------------------------------------------------------------------

Perhaps the "numpy.ndarray size changed" error might be relevant?

> I have located my rpi, now I need to put an OS on it and wire it up
> somewhere.

That sounds like devotion beyond the call of duty.  Thanks. :)

-- 

73,
Ged.

**Comment by totaam** [comment]
Here is the full log of what I did:
* downloaded `2021-05-07-raspios-buster-armhf-full.zip` and extracted it to an sd card
* booted this OS on:
```
$ cat /sys/firmware/devicetree/base/model
Raspberry Pi 4 Model B Rev 1.1
```
* changed password, logged in via ssh, then `apt-get install` lots of packages as per https://github.com/Xpra-org/xpra/blob/master/docs/Build/Debian.md
* `git clone https://github.com/Xpra-org/xpra`
* `cd xpra;python3 ./setup.py build;sudo python3 ./setup.py install --prefix=/usr`
* `PYTHONPATH=/usr/lib/python3.7/site-packages/ xpra start --start=xterm --no-daemon`
* from an xterm on the default X11 desktop: `PYTHONPATH=/usr/lib/python3.7/site-packages/ xpra attach`

The `xterm` appeared after a few seconds and worked fine.
I also tried connecting to remote hosts via tcp, also without problems.

For good measure, I then dusted off an older rpi, downloaded an older version of raspios since newer versions fail to boot on it.
```
$ cat /sys/firmware/devicetree/base/model
Raspberry Pi Model B Rev 2
```
That took **many many** hours longer.
And also required to `git checkout v3.1.x` before building because the Python3 version in Debian Stretch (3.5.x) is just too old for newer versions of xpra.
With this one, I do get the black window bug. There is one obvious warning sticking out every time there's a missing screen update: `Couldn't find foreign struct converter for 'cairo.Context'`.
Easily fixed by installing the cairo bindings: `sudo apt-get install python3-gi-cairo`.
This dependency is usually installed automatically when installing packages: https://github.com/Xpra-org/xpra/blob/c787b6465ed9a161e5e0bdbd5f59ca4daa89310b/packaging/debian/xpra/control#L45-L46
I'll update the build instructions to include this package as part of the build preparation (even though it's not actually needed for building..)

See also #3291.
